# Coalescent Simulator for Tumor Evolution (CSiTE)

CSiTE is a Python program for simulating short reads for tumor samples.
It takes a coalescent tree (in the newick format) and simulates Single Nucleotide Variants
(SNVs) and Copy Number Variants (CNVs) along the history of the genealogy. 
By integrating those somatic variants into the normal genome, perturbed 
genome of each tumor cell can be built, which are then piped to ART to simulate NGS short reads. 
Germline variants can also be integrated to the genome to make the data more realistic. 
CSiTE can be used for both bulk tumor and single cell data simulation.

## Installing

CSiTE is written in Python3 (>=3.5), and requires numpy, pyfaidx and yaml. 
[ART](https://www.niehs.nih.gov/research/resources/software/biostatistics/art/index.cfm) 
is also required if you want to simulate short reads with CSiTE.

CSiTE can be downloaded from github:

    git clone https://github.com/hchyang/CSiTE.git

## Running CSiTE

There are four modules in CSiTE. 

    Program: csite.py (a Coalescent Simulator for Tumor Evolution)
    Version: 0.9.0

    Usage:   csite.py <command> [options]

    Command: vcf2fa      integrate germline snp
             phylovar    simulate somatic vars
             draft2fa    build reference
             quaternity  a wrapper for the whole pipeline

### Module vcf2fa 

The `vcf2fa` module can be used to build the genome of normal cells. It takes a 
reference genome (e.g. hg19.fa) and a list of SNPs in VCF format as input (only
SNPs are acceptable). The variants in VCF should be phased. The sequence of each 
haplotype, which contains the variants on each them, will be output.

### Module phylovar

This module the the core module of CSiTE.

To run the program, a coalescent tree which captures the ancestral relationships
of the tumor cells in a sample is required. The input tree file should be in the
newick format, which can be generated by the [ms
program](http://home.uchicago.edu/rhudson1/source/mksamples.html) with the `-T`
option. ms program has the full-assemblage of the options needed to generate
complex demographic histories of the sample. 

The two most important parameters in the simulation is the mutation rate of SNVs
(`--snv_rate`) and CNVs (`--cnv_rate`). The mutational events in CSiTE are
simulated according to the Poisson process with user specified parameters (see
[Notes](https://github.com/hchyang/CSiTE#notes) for extra discussions).  

Other than the rate of CNVs, there are five other parameters guiding CNVs
simulation. 

`--cnv_length_beta` can be used to specify the beta/mean parameter of the
exponential distribution (We assume the length of CNVs follows an exponential
distribution. This parameter specifies the mean length of the CNV events).

`--cnv_length_max` can be used to set the upper limit of CNVs' length (This will
effectively truncate the exponential distribution at this limit). 

`--del_prob` A CNV event can be a deletion or an amplification. Using this
parameter, we can specify the probability that a CNV event is a deletion.

`--copy_max` can be used to set the upper bound of an amplification event.  When
an amplification event happens, it will randomly pick a copy number according to
a geometric like distribution with Pr(n+1)=p\*Pr(n). The p parameter can be
specified by `-c/--copy_parameter`. The overall distribution will be normalized
s.t. the total probability is 1. 

The coalescent tree only describes the polymorphic part of the sample. You can
also simulate the truncal part of the tumor evolution (tumorigenesis) through
two different approaches: a) specify the trunk length e.g. using
`--trunk_length`. `--trunk_length` accepts a single float number that can be
used as the branch length leading to the root of the coalescent tree.  b)
specify the events on the trunk explicitly in a file through `--trunk_vars
filename`. The format of the trunk variants file is described in the later
section.

When the number of simulated cells is very large, the computational load can be
very heavy. Given the fact that most of the mutational events are extremely
rare, we implemented a pruning algorithm to trim the tree using `--prune` and
`--prune_proportion` options. For example, if you want to simulate the somatic
variants for a population containing 1,000,000 cells, after you setting `--prune
100` or `--prune_proportion 0.0001`, all subtrees with <=100 tips will be
trimmed into a tip node, which means there will be no polymorphic variants on
those subtrees with <=100 tips. So the tips belonging to the same subtree (with
<=100 tips) will show the same genotypes.

Other convenient options including:
- -D/--depth to simulate the variants under different sequencing coverage (i.e.
  given the specified coverage d, we will simulate a sequence coverage of x,
  where x is sampled from Poission(d)). 
- -p/--ploidy to set the ploidy of tumor cells to simulate
- -P/--purity to set the purity of tumor sample
- --length to set the length of sequence to simulate

###### Notes
By default, the branch length from the ms program is measured in 4N
generations. CSiTE will simulate SNVs/CNVs events following the Poisson process.
For example, if we set -r 100, this is equivalent to set the population rescaled
parameter θ=4Nu as 100 (see ms manual for details). 

#### Input files

##### Tree file (-t/--tree)

Tree file should be in newick format, for example:

    ((2:0.083,4:0.083):0.345,(5:0.322,(1:0.030,3:0.030):0.292):0.105);

##### Trunk variants file (--trunk\_vars)

You can simulate truncal variants by specify `--trunk_length`, or import trunk
variants directly using `--trunk_vars` option.  The file format of truncal
variants is like:
    
    #parental start end copy
    0 1 1 0
    1 12 123 +3 
    0 12 123 -1

- **parental**:  which parental copy the variant locates in (0 means one of the
  parental copy, 1 means the other copy)
- **start**:     the start position of the variant
- **end**:       the end position of the variant
- **copy**:      an integer. 0: SNV, -1: deletion, +int: amplification

P.S. start and end are 0 based. And the region of each variant is similar to the
bed file (inclusive on the left, but exclusive on the right end,
i.e.\[start,end)).

#### Output files

##### SNVs file (-S/--snv)

This file contains the frequency information of simulated SNVs. It contains four
columns. 
- **position**:        the position (0-based) of SNVs
- **true\_freq**:      the true frequency of the alternative allele in the
  sample. 
- **total\_depth**:    the simulated total coverage of tumor and normal cells at
  the position of SNV. 
- **simulated\_freq**: the observed frequency of alternative allele across tumor
  and normal cell population.

##### Genotype file (--snv\_genotype)

This file contains the snv\_genotype of each tumor cell at SNV loci. Each SNV
has one record. The first column is the coordinate of the SNV. Subsequently,
there is one column for each tumor cell. The snv\_genotype is in the form of
‘M:N’. M denotes the number of alternative allele and N denotes the number of
reference allele.

##### Individual CNV file (--ind\_cnvs)

This file contains the CNVs on each parental copy of each single cell in the
sample. There are five columns in this file:
    
    #cell parental  start     end       copy
    1       0       7912422   7930111   2
    1       1       43110140  43341629  1
    2       0       2255734   2299608   -1
    2       0       22660687  22788472  -1
    2       1       59756841  61142076  3

- **cell**:     the id of each cell in the sample
- **parental**: which parental copy the variant locates in (0 means one of the
  parental copy, 1 means another copy, 2 means the third copy if your sample is
  triploid...)
- **start**:    the start position of the CNV
- **end**:      the end position of the CNV
- **copy**:     an integer. -1: deletion; positive integer: amplification

P.S. start and end are 0 based. And the region of each variant is similar to the
bed file (inclusive on the left, but exclusive on the right end,
i.e.\[start,end)).

##### Named tree file (-T/--named\_tree)

CSiTE can output a [NHX
file](https://sites.google.com/site/cmzmasek/home/software/forester/nhx) with
each node named. Combined with nodes variants file below, users can fingure out
the locations of SNVs/CNVs on the tree.  We name each internal node with the
format 'Node:XXX', XXX is an integer and start from N+1, where N is the number
of leaves in the input tree. And 'Node:N+1' is the name of the root node. The
names of each leaf will be kept as the input tree.

##### Nodes variants file (-n/--nodes\_vars)

CSiTE can output the variants (SNVs/CNVs) occured on the branch leading to each
node. There are five columns in this file:

    #node parental        start   end     copy 
    ...
    ...
    ...

- **node**:     the id of nodes or leaves in the sample
- **parental**: which parental copy the variant locates in (0 means one of the
  parental copy, 1 means another copy, 2 means the third copy if your sample is
  triploid...)
- **start**:    the start position of the variant
- **end**:      the end position of the variant
- **copy**:     an integer. 0: SNV; -1: deletion; positive integer:
  amplification

##### CNV profile file (--cnv\_profile)

This file contains the CNV profile across the whole segment. There are 3 columns
in this file. The first column is the start of the CNV region, and the second is
the end of the CNV region. Both breakpoints are 0-based, defined CNVs covered on
\[start, end). The last column is the total copies of each region.

##### parental copy file (--parental\_copy)

This file contains the information of parental copy for each SNV. The first
column is the coordinate of the SNV, and followed by N columns if the ploidy is
N.

##### Log file (-g/--log)

This file contains logging information, e.g. the command line parameters and the
random seed used. You can use these information to replicate the simulation.
After setting the `--loglevel` as DEBUG, CSiTE will output detailed information
of the simulation process. 

#### draft file (--draft)

This file contains the information of perturbed genome of simulated tumor cell.

### Module draft2fa

After generating genomes of normal cells with `vcf2fa` and genome draft of tumor
cells with `phylovar`, we can build the genome sequences for each tumor sample. 
This work is done by the module `draft2fa`. And the genomes of tumor cells and 
normal cells can then be used by ART to simulate short reads.

### Module quaternity

This is a convenient wrapper for simulating reads for tumor sample. It calls the
module `vcf2fa` to build the genome of normal cells, and calls the modules `phylovar`
and `draft2fa` to build the genomes of tumor cells. And then according the settings
customised by user, `quaternity` calls ART to simulte short reads for them.

### Examples

* Simulate the coalescent tree of a sample of 1000 tumor cells, which are
sampled from a exponetially growing tumor. 
(consult the manual of ms for more information)

    `ms 1000 1 -T -G 1 |tail -n1 > ms_tree.txt`

* Simulate the somatic SNVs of this sample. We assume the sequencing depth is
60, and the purity of the sample is 0.8, which means there are 250 normal
cells other than these 1000 tumor cells. Other settings are: 
a) the mutation rate of SNVs and CNVs are 10 and 0.1 respectively; 
b) the sequence length is 135534747 (chr10 of hg19); 
c) the cells of the sample are diploid. We save the
frequncy of the simulated SNVs to file 'snvs\_freq.txt'.

    `csite.py -t ms_tree.txt -P 0.8 --length 135534747 -r 10 -R 0.1 -D 60 -S
    snvs_freq.txt`

* There are no truncal muations in the simulation above. If you want to simulate
truncal muations, use the option `--trunk_length`.

    `csite.py -t ms_tree.txt -P 0.8 --length 135534747 -r 10 -R 0.1 -D 60 -S
    snvs_freq.txt --trunk_length 2.0`

* If you want to ignore the variants with the frequency <=0.01, you can use
`--prune 20` or `--prune_proportion 0.02` (we use `--prune 20` instead of
`--prune 10` for the cells are diploid in the our simulation). These two
options can be used to accelerate the simulation when your tree is huge.

    `csite.py -t ms_tree.txt -P 0.8 --length 135534747 -r 10 -R 0.1 -D 60 -S
    snvs_freq.txt --trunk_length 2.0 --prune 20`

    or

    `csite.py -t ms_tree.txt -P 0.8 --length 135534747 -r 10 -R 0.1 -D 60 -S
    snvs_freq.txt --trunk_length 2.0 --prune_proportion 0.02`

* If you want to save the SNVs genotypes for each single cell for exactly the
same simulation as above, use the options `--snv_genotype` and
`--random_seed`.

    `csite.py -t ms_tree.txt -P 0.8 --length 135534747 -r 10 -R 0.1 -D 60 -S
    snvs_freq.txt --trunk_length 2.0 --prune_proportion 0.02 --snv_genotype
    snvs_genotype.txt --random_seed xxxx`

    P.S. The random seed xxxx can be found in the log file of the previous
    simulation.

## Authors

* [Hechuan Yang](https://github.com/hchyang)

## License

This project is licensed under the GNU GPLv3 License - see the
[LICENSE](LICENSE) file for details.

